docs/build-javascaladocs.sh

```diff
@@ -1,23 +1,58 @@
-#! /bin/bash
-
-VERSION=$1
-
-echo $VERSION
-
-if [[ -z "${VERSION}" ]]; then
-  echo "Version not defined"
-else
-  (cd target && \
-    rm -rf cloudflow && \
-    git clone --depth=100 --branch "$VERSION" https://github.com/lightbend/cloudflow.git && \
-    cd cloudflow && \
-    export DIR=$(yq e '.version' docs/docs-source/docs/antora.yml) && \
-    echo $DIR && \
-    cd core && \
-    (sbt -mem 2048 clean unidoc || true) && \
-    cd ../../../ && \
-    ls -al "./target/cloudflow/core/target" && \
-    ls -al "./target/cloudflow/core/target/scala-2.12" && \
-    mv "./target/cloudflow/core/target/scala-2.12/unidoc" "./target/staging/docs/$DIR/api/scaladoc" && \
-    mv "./target/cloudflow/core/target/javaunidoc" "./target/staging/docs/$DIR/api/javadoc")
+#!/usr/bin/env bash
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+ROOT_DIR=$( cd "${SCRIPT_DIR}/.." && pwd )
+DOCS_DIR="$ROOT_DIR/docs"
+
+# Keep this consistent with the `work_dir` in the Makefile.
+TARGET_DIR="$DOCS_DIR/target"
+CLOUDFLOW_CLONE_DIR="${TARGET_DIR}/cloudflow"
+
+echo "SCRIPT_DIR          => $SCRIPT_DIR"
+echo "ROOT_DIR            => $ROOT_DIR"
+echo "DOCS_DIR            => $DOCS_DIR"
+echo "TARGET_DIR          => $TARGET_DIR"
+echo "CLOUDFLOW_CLONE_DIR => $CLOUDFLOW_CLONE_DIR"
+
+versions=$(yq e '.content.sources[0].branches[]' "$DOCS_DIR/docs-source/site.yml")
+
+if [ -z "$versions" ]; then
+  echo "No version branches found in $DOCS_DIR/docs-source/site.yml. See file content below:"
+  cat "$DOCS_DIR/docs-source/site.yml"
+  exit 1
+fi
+
+# Recreating the target dir before generating the new docs.
+rm -rf "$TARGET_DIR/cloudflow"
+mkdir -p -v "$TARGET_DIR"
+
+git clone https://github.com/lightbend/cloudflow.git "$CLOUDFLOW_CLONE_DIR"
+
+FAILURES=()
+
+for version in $versions; do
+  echo "Building docs for version $version"
+  cd "$CLOUDFLOW_CLONE_DIR" || exit 1 # Maybe if failed to clone the cloudflow repo.
+  git checkout --quiet "$version"
+  cd "$CLOUDFLOW_CLONE_DIR/core" || exit 1 # core directory does not exists anymore?
+
+  if ! sbt --error --supershell=false -mem 2048 clean unidoc; then
+    FAILURES+=("$version")
+  fi
+
+  ls -al "$CLOUDFLOW_CLONE_DIR/core/target"
+  ls -al "$CLOUDFLOW_CLONE_DIR/core/target/scala-2.12"
+
+  DIR=$(yq e '.version' "$CLOUDFLOW_CLONE_DIR/docs/docs-source/docs/antora.yml")
+  mkdir -p "$TARGET_DIR/staging/docs/$DIR/api"
+
+  mv -v "$CLOUDFLOW_CLONE_DIR/core/target/scala-2.12/unidoc" "$TARGET_DIR/staging/docs/$DIR/api/scaladoc"
+  mv -v "$CLOUDFLOW_CLONE_DIR/core/target/javaunidoc"        "$TARGET_DIR/staging/docs/$DIR/api/javadoc"
+done
+
+# So that we can at least log the errors, but it won't fail the script.
+if [ ${#FAILURES[@]} -gt 0 ]; then
+  printf "Failed to generate docs for version %s\n" "${FAILURES[@]}"
+  cd "$ROOT_DIR" || true # get back home
+  exit 0
```

Comment:
Should we fail the script? I guess it is better than keeping the publishing process going and fail silently.

Reply:
> the publishing process going and fail silently

this was done (sadly) on purpose, some of the old versions of Cloudflow return an error code from the `sbt unidoc` command despite producing the actual documentation.

Now we removed some old versions and it might work, have you tested it in CI @marcospereira ?

Reply:
Yeah, I would say let's remove these old versions since they are not even maintained, right?

> have you tested it in CI @marcospereira ?

No. Shall we remove the old-failing versions before doing that?

Reply:
> No. Shall we remove the old-failing versions before doing that?

I don't know if we should remove version e.g. `1.3.3` can you check if/on which versions the script is failing?